packages/compass-components/src/hooks/use-confirmation.tsx

```diff
@@ -0,0 +1,147 @@
+
+export { ConfirmationModalVariant };
+
+type ConfirmationProperties = {
+  title: string;
+  description: string;
+  buttonText?: string;
+  variant?: ConfirmationModalVariant;
+};
+type ConfirmationCallback = (value: boolean) => void;
+
+interface ConfirmationModalContextData {
+  showConfirmation: (
+    props: ConfirmationProperties,
+    callback: ConfirmationCallback
+  ) => void;
+  isMounted: boolean;
+}
+
+const ConfirmationModalContext =
+  React.createContext<ConfirmationModalContextData>({
+    isMounted: false,
+  } as ConfirmationModalContextData);
+
+class ShowConfirmationEvent extends Event {
+  constructor(
+    public props: ConfirmationProperties,
+    public callback: ConfirmationCallback
+  ) {
+    super('show-confirmation');
+  }
+}
+class GlobalConfirmation extends EventTarget {
+  showConfirmation(props: ConfirmationProperties) {
+    return new Promise<boolean>((resolve) => {
+      const event = new ShowConfirmationEvent(props, resolve);
+      this.dispatchEvent(event);
+    });
+  }
+}
+const globalConfirmation = new GlobalConfirmation();
+export const showConfirmation =
+  globalConfirmation.showConfirmation.bind(globalConfirmation);
+
+type ConfirmationModalAreaProps = ConfirmationProperties & {
+  open: boolean;
+};
+
+export const ConfirmationModalArea: React.FC = ({ children }) => {
+  const hasParentContext = useContext(ConfirmationModalContext).isMounted;
+
+  const [confirmationProps, setConfirmationProps] = useState({
+    open: false,
+  } as ConfirmationModalAreaProps);
+  const callbackRef = useRef<(value: boolean) => void>();
+
+  useEffect(() => {
+    return () => {
+      callbackRef.current?.(false);
+    };
+  }, []);
+
+  // Event listener to use confirmation modal outside of react
+  useEffect(() => {
+    const listener = ({ props, callback }: ShowConfirmationEvent) => {
+      showConfirmation(props, callback);
+    };
+    globalConfirmation.addEventListener(
+      'show-confirmation',
+      listener as EventListener
+    );
+    return () => {
+      globalConfirmation.removeEventListener(
+        'show-confirmation',
+        listener as EventListener
+      );
+    };
+  }, []);
+
+  const showConfirmation = (
+    props: ConfirmationProperties,
+    callback: ConfirmationCallback
+  ) => {
+    if (confirmationProps.open) {
+      throw new Error('Confirmation modal is already open');
+    }
+    setConfirmationProps({ open: true, ...props });
+    callbackRef.current = callback;
+  };
+
+  const handleConfirm = () => {
+    onUserAction(true);
+  };
+
+  const handleCancel = () => {
+    onUserAction(false);
+  };
+
+  const onUserAction = (value: boolean) => {
+    setConfirmationProps((state) => ({ ...state, open: false }));
+    callbackRef.current?.(value);
+    callbackRef.current = undefined;
+  };
+
+  if (hasParentContext) {
+    return <>{children}</>;
+  }
+
+  return (
+    <ConfirmationModalContext.Provider
+      value={{ showConfirmation, isMounted: true }}
```

Comment:
As I [said before](https://github.com/mongodb-js/compass/pull/4079#discussion_r1113046863), the value passed to the provider should be memoized to prevent components depending on this context from re-renders, this is important for context providers as context access goes even through things like React.memo and can cause unexpected re-renders